src/sink/mod.rs

```diff
@@ -28,42 +29,48 @@ pub use self::null::{Null, NullConfig};
 pub use self::prometheus::{Prometheus, PrometheusConfig};
 pub use self::wavefront::{Wavefront, WavefrontConfig};
 
-/// A 'sink' is a sink for metrics.
-pub trait Sink {
-    /// Lookup the `Sink`'s specific flush interval. This determines how often a
-    /// sink will obey the periodic flush pulse.
-    ///
-    /// If the value is `None` this is a signal that the sink will NEVER flush
-    /// EXCEPT in the case where the sink's valve_state is Closed.
-    fn flush_interval(&self) -> Option<u64>;
-    /// Perform the `Sink` specific flush. The rate at which this occurs is
-    /// determined by the global `flush_interval` or the sink specific flush
-    /// interval. Pulses occur at a rate of once per second, subject to
-    /// communication delays in the routing topology.
-    fn flush(&mut self) -> ();
-    /// Lookup the `Sink` valve state. See `Valve` documentation for more
-    /// information.
-    fn valve_state(&self) -> Valve;
-    /// Deliver a `Telemetry` to the `Sink`. Exact behaviour varies by
-    /// implementation.
-    fn deliver(&mut self, point: sync::Arc<Option<Telemetry>>) -> ();
-    /// Deliver a `LogLine` to the `Sink`. Exact behaviour varies by
-    /// implementation.
-    fn deliver_line(&mut self, line: sync::Arc<Option<LogLine>>) -> ();
-    /// Deliver a 'Raw' series of encoded bytes to the sink.
-    fn deliver_raw(&mut self, _encoding: Encoding, _bytes: Vec<u8>) -> () {
-        // Not all sinks accept raw events.  By default, we do nothing.
-        return;
+
+/// Wrapper around a given sink implementation and its config.
+pub struct StatefulSink<S, SConfig>
+    where
+        S: Send + Sync + Sink<SConfig>,
+        SConfig: 'static + Send + Sync + Clone,
+{
+    recv: hopper::Receiver<Event>,
+    sources: Vec<String>,
+    #[allow(dead_code)] // Only here to avoid using PhantomData.
+    config: SConfig,
```

Comment:
Maybe I'm missing it, but I can't find a place where config is being used by StatefulSink?

If there's only the one use site you might be able to remove the clone, since only Sink would own its config.